README.md
yatc
====

Yet Another Thin Client - small gui application to start freerdp session to MS Terminal Server.

YATC can use FreeRDP or "2X RDP Client" as backend for RDP connection.


Changelog
YATC 0.9.11:
  yatc.py:
    1. Mounter improvements: check device is connected through USB even if it isn't 'removable'; check that device is storage device.

  post-install.sh:
    1. Re-design 2xclient installation.


  uprgade.sh:
    1. Re-design 2xclient installation.

YATC 0.9.10:
  yatc.py:
    1. Add 'utf8=1' option for mount.

  post-install.sh:
    1. Add extra (Russian) locale.

  uprgade.sh:
    1. Add extra (Russian) locale.

YATC 0.9.9:
  yatc.py:
    1. Mounter class added.

  post-install.sh:
    1. Switch from autofs to mounter.
    2. Link to 2xclient changed from latest to particular version.

  uprgade.sh:
    1. 2xclient upgrade added.

YATC 0.9.8:
  yatc.py:
    1. Merge dev to master.

  post-install.sh:
    1. Merge dev to master.
    2. Updating 2xclient actual version to 14.0.3213.

YATC 0.9.7:
  yatc.py:
    1. Switched from ugly error windows to pretty informational frame. It's placed right above login frame.
    2. BUG fixed: 2xclient - can't establish new connection after unsuccessful connection without reboot.
      The reason was duplication  of connection parameters in dictionary.
    3. Some new RDP errors exceptions added.

YATC 0.9.6:
  yatc.py:
    1. "Server idle timeout reached" exception handling added.

YATC 0.9.5:
  yatc.py:
    1. Stripping of string entered to login and password fields added.
    2. BUG fixed: freerdp - can't establish connection again without reboot.
      The reason was duplication  of connection parameters in dictionary.
    3. Added "Error" dialog windows after unsuccessfull RDP connection.

  post-install.sh:
    1. Amixer options changed to make Master and PCM sound louder (100%).

YATC 0.9.4:
  yatc.py:
    1. Add option to shutdown thin client if RDP not connected for some time.

YATC 0.9.3:
  post-install.sh:
    1. 2xclient source changed to vendor site URL.
    2. Added 2xclient version check. If version changed Changelog will be displayed. 
    3. Some output sugar added. 

YATC 0.9.2:
  yatc.py:
    1. Add support of multiple RDP backends: FreeRDP or "2X RDP client".
      Now freerdp or 2xclient argument must be specified at startup.

  post-install.sh:
    1. Add parameter to choose RDP backend.
      Now freerdp or 2xclient argument must be specified at startup.

YATC 0.9.1:
  yatc.py:
    1. Add support of "2X RDP client" as backend for RDP connection.

  post-install.sh:
    1. Add support of "2X RDP client" as backend for RDP connection.
YATC 0.8.4:
  yatc.py:
    1. Print version to  ~/.yatc/version file on startup.

  post-install.sh:
    1. Changing FreeRDP source version.

YATC 0.8.3:
  yatc.py:
    1. Added dialog to re-ask user about reboot and shutdown.
    2. Exception handling for freerdp improved. Will log Freerdp exit code and output now.
    3. Write version to log.
    4. Write version to STDOUT if '-v' command line argument specified.

  post-install.sh:
    1. YATC_REPO variable added.
    2. YATC_BRANCH switched to 'master'.

  uprgade.sh:
    1. Script for upgrade added.
    2. YATC_BRANCH switched to 'master'.

YATC 0.8.2:
  post-install.sh:
    1. Forked FreeRDP to nixargh account to freeze current state. 

YATC 0.8.1:
  yatc.py:
    1. New log path - "~/.yatc/yatc.log".
    2. New config path - "~/.yatc/yatc.conf".
    3. Crypt class moved to separate file.

  post-install.sh:
    1. YATC now has it own configuration directory under user home.
    2. YATC binary path changed.
    3. YATC lib directory added.
    4. FreeRDP branch switched to "stable-1.1".

    
YATC 0.7.5:
  post-install.sh:
    1. Adding "or" to not stop if alsa doesn't have PCM device.


YATC 0.7.4:
  post-install.sh:
    1. Fixing pulse setup.


YATC 0.7.3:
  post-install.sh:
    1. More requirements for pulseaudio added.
    2. Pulseaudio access groups membership for user added.
    3. pulseaudio daemon now start at "user" session.
    4. Unmuting alsa Master and PCM + pulseaudio device with index "0".


YATC 0.7.2:
  post-install.sh:
    1. Fixed user group from sound to audio.


YATC 0.7.1:
  yatc.py:
    1. Alsa replaced by pulseaudio.
    
  post-install.sh:
    1. Alsa removed from dependencies.
    2. libpulse-dev added to dependencies.


YATC 0.7.0:
  yatc.py:
    1. CDROM redirection added.

  post-install.sh:
    1. usbmount replaced by autofs.
    2. YATC_BRANCH variable added.
    3. Alsa added to dependencies.
    4. "user" added to sound group.


YATC 0.6.1:
  yatc.py:
    1. Sound forwarding option added.

  post-install.sh:
    1. "set -u -e" added. Few sed command lines fixed.


sysconf/steps.sh
# must be executed from root
##### Settings ################################################################
USER=user
ADM_USER=admuser
###############################################################################

# install packages
apt-get update
apt-get install -y python-software-properties
apt-add-repository -y ppa:freerdp-team/freerdp
apt-get update
apt-get install -y freerdp-x11 python3 git xorg python3-tk vim cups puppet

# create user
useradd -m -U -c "RDP User" -G shadow -s /bin/bash $USER

# create user for remote configuration
#useradd -m -U -G sudo -c "Remote configuration" commander

# add group for ssh access
groupadd ssh_users

# give ssh access to Metrex User and Remote configuration user
usermod -a -G ssh_users $ADM_USER

# configure ssh
SSHD_CONF=/etc/ssh/sshd_config
sed -i {s/PermitRootLogin yes/PermitRootLogin no/} $SSHD_CONF 
echo -e "AllowGroups\tssh_users" >> $SSHD_CONF
service ssh restart

# install python module for pam authentication
cd /home/$USER
git clone https://github.com/leonnnn/python3-simplepam.git
cd ./python3-simplepam
python3 setup.py install

# install YATC
cd /home/$USER
git clone https://github.com/nixargh/yatc.git

# make user autologin
# http://blog.shvetsov.com/2010/09/auto-login-ubuntu-user-from-cli.html
TTY_CONF=/etc/init/tty1.conf
sed -i {s/exec/#exec/} $TTY_CONF 
echo "exec /bin/login -f $USER < /dev/tty1 > /dev/tty1 2>&1" >> $TTY_CONF

# add some rights to user
echo -e "$USER\tALL=(root) NOPASSWD:/sbin/reboot,/sbin/poweroff\n" >> /etc/sudoers

# setup X and application to start on boot at terminal
#
# /home/user/.bash_profile
BASH_PROFILE=/home/$USER/.bash_profile
echo ". startx" > $BASH_PROFILE

# /home/user/.xinitrc
XINITRC=/home/$USER/.xinitrc
echo "./yatc/yatc.py  -- -depth 16" > $XINITRC

# fix onership
chown $USER:$USER /home/$USER

# set cupsd to listen on all interfaces
# and allow access from subnet
CUPSD_CONF=/etc/cups/cupsd.conf
mv $CUPSD_CONF $CUPSD_CONF.orig
awk '/\/Location/ {print "  Allow from 10.0.*"; print; next }1' $CUPSD_CONF.orig >> $CUPSD_CONF
sed -i {s/127.0.0.1/0.0.0.0/} $CUPSD_CONF
service cups restart


sysconf/upgrade.sh
#!/bin/bash
# script to upgrade YATC on Ubuntu 12.04 - 14.04
# (*w) author: nixargh <nixargh@gmail.com>
VERSION="0.9.11"
##### Settings ################################################################
# !!! must be executed from root !!!
RDPUSER="user"
FREERDP_REPO="https://github.com/FreeRDP/FreeRDP.git"
FREERDP_BRANCH="c9bc88d5f0fed0de03ee697dd382ba8f8a434a82"
YATC_REPO="https://github.com/nixargh/yatc.git"
YATC_BRANCH="master"
TMP_DIR="/tmp"
TWOXCLIENT_VER="14.0.3213"
TWOXCLIENT="http://www.2x.com/downloads/builds/applicationserver/${TWOXCLIENT_VER}/2XClient.deb"
LOCALE="ru_RU.UTF-8"
###############################################################################
set -u -e

# Check that you are root.
#
check_root() {
  if [ $USER != "root" ]; then
    echo "You must be root."
    return 1
  fi
  return 0
}

# Upgrade FreeRDP.
#
upgrade_freerdp() {
  cd $TMP_DIR
  git clone $FREERDP_REPO
  cd ./FreeRDP
  git checkout $FREERDP_BRANCH

  cmake -DCMAKE_BUILD_TYPE=Debug -DWITH_SSE2=ON -DWITH_CUPS=ON -DCHANNEL_PRINTER=ON -DWITH_PULSE=ON .
  make
  make install

  return 0
}

# Upgrade YATC.
#
upgrade_yatc() {
  local YATCBIN=/usr/bin/yatc
  local YATCLIB=/usr/lib/yatc

  cd $TMP_DIR
  git clone $YATC_REPO
  cd ./yatc
  git checkout $YATC_BRANCH

  python3 -mpy_compile ./yatc.py
  mv ./__pycache__/yatc.*.pyc $YATCBIN
  chmod 755 $YATCBIN

  if [ $(dpkg -l |grep autofs |wc -l) -gt 0 ]; then
    apt-get purge -y autofs
  fi

  # add mount, umount to sudoers
  sed -i "s/$RDPUSER\tALL=(root) NOPASSWD:.*//g" /etc/sudoers
  echo -e "$RDPUSER\tALL=(root) NOPASSWD:/sbin/reboot,/sbin/poweroff,/bin/mount,/bin/umount\n" >> /etc/sudoers

  # create directory to store mountpoints
  test -d /media/usbdisk || mkdir /media/usbdisk
  chown $RDPUSER /media/usbdisk

  # Add extra locale
  locale-gen $LOCALE
  update-locale

  return 0
}

# Upgrade 2xclient.
#
upgrade_twoxclient() {
	echo -e "\t2X RDP client upgrade starting.\n"

	dpkg --add-architecture i386

	cd /tmp
	wget $TWOXCLIENT -O ./2xclient.deb

	dpkg -i ./2xclient.deb || apt-get -f -y install
	dpkg -i ./2xclient.deb

	echo -e "\t2X RDP client upgrade finished."
	return 0
}

# Create user config directory.
#
create_config_dir() {
  local CONF_DIR="/home/$RDPUSER/.yatc"

  if [ ! -d $CONF_DIR ]; then
    mkdir $CONF_DIR
    chown $RDPUSER:$RDPUSER $CONF_DIR
  fi
  return 0
}

# Unmute alsa & pulseaudio.
#
unmute() {
  if [ `ps aux | grep -v grep |grep -c pulseaudio` -eq 0 ]; then
    for DEV in "PCM" "Master"; do
      amixer set $DEV 100 unmute || echo "Can't unmute $DEV. Skipping..."
    done
    sudo -i -u $RDPUSER pulseaudio -D
    sleep 1
    sudo -i -u $RDPUSER pactl set-sink-mute 0 0
  fi

  return 0
}
##### BEGIN ###################################################################

JOB=$1

echo "Starting upgrade of $1."

check_root || exit 0

case $JOB in
  freerdp)
    upgrade_freerdp
  ;;
  2xclient)
    upgrade_twoxclient
  ;;
  yatc)
    create_config_dir
    upgrade_yatc
  ;;
esac

# fix onership
chown $RDPUSER:$RDPUSER -R /home/$RDPUSER

echo "Upgraded."


sysconf/post-install.sh
#!/bin/bash
# script to deploy YATC on Ubuntu 12.04 - 14.04 (netinstall with only ssh server installed)
# (*w) author: nixargh <nixargh@gmail.com>
VERSION="0.9.11"
##### Settings ################################################################
# !!! must be executed from root !!!
RDPUSER="user"
TIMEZONE="Europe/Moscow"
FREERDP_REPO="https://github.com/FreeRDP/FreeRDP.git"
FREERDP_BRANCH="c9bc88d5f0fed0de03ee697dd382ba8f8a434a82"
YATC_REPO="https://github.com/nixargh/yatc.git"
YATC_BRANCH="master"
TWOXCLIENT_VER="14.0.3213"
TWOXCLIENT="http://www.2x.com/downloads/builds/applicationserver/${TWOXCLIENT_VER}/2XClient.deb"
LOCALE="ru_RU.UTF-8"
###############################################################################
set -u -e

# check that you are root
check_user() {
  if [ $USER != "root" ]; then
    echo "You must be root."
    exit 1
  fi
}

common() {
  echo -e "\tCommon configuration starting.\n"

  # set timezone
  cp -f /usr/share/zoneinfo/$TIMEZONE /etc/localtime

  DATE1=`date +%s%N`

  # detect admuser
  ADM_USER=`grep 1000 /etc/passwd |awk 'BEGIN{FS=":"} {print $1}'`

  # install required packages
  apt-get update
  apt-get install -y python3 python3-tk python3-crypto git xorg vim cups puppet libasound2 \
    libasound2-plugins libasound2-plugins:i386 alsa-utils alsa-oss pulseaudio pulseaudio-utils dbus-x11 curl

  # create user
  useradd -m -U -c "RDP User" -G shadow,audio,pulse,pulse-access -s /bin/bash $RDPUSER

  # create user config directory
  CONF_DIR="/home/$RDPUSER/.yatc"
  mkdir $CONF_DIR
  chown $RDPUSER:$RDPUSER $CONF_DIR

  # add group for ssh access
  groupadd ssh_users

  # give ssh access to Metrex User and Remote configuration user
  usermod -a -G ssh_users $ADM_USER

  # configure ssh
  SSHD_CONF=/etc/ssh/sshd_config
  sed -i "{s/PermitRootLogin yes/PermitRootLogin no/}" $SSHD_CONF 
  echo -e "AllowGroups\tssh_users" >> $SSHD_CONF
  service ssh restart

  # Add extra locale
  locale-gen $LOCALE
  update-locale

  # Unmute alsa & pulseaudio
  for DEV in "PCM" "Master"; do
	  amixer set $DEV 100 unmute || echo "Can't unmute $DEV. Skipping..."
  done
  sudo -i -u $RDPUSER dbus-launch --exit-with-session pulseaudio --daemon
  sleep 1
  sudo -i -u $RDPUSER pactl set-sink-mute 0 0

  # install python library for pam authentication
  cd /tmp
  git clone https://github.com/leonnnn/python3-simplepam.git
  cd ./python3-simplepam
  python3 setup.py install

  # install python library for simple encryption and decryption
  cd /tmp
  git clone https://github.com/andrewcooke/simple-crypt.git
  cd ./simple-crypt/
  python3 setup.py install

  # install YATC
  YATCBIN=/usr/bin/yatc
  YATCLIB=/usr/lib/yatc
  cd /tmp
  git clone $YATC_REPO
  cd ./yatc
  git checkout $YATC_BRANCH

  DATE2=`date +%s%N`

  python3 -mpy_compile ./yatc.py
  mv ./__pycache__/yatc.*.pyc $YATCBIN
  chmod 755 $YATCBIN

  CRYPT_LIB="./lib/crypt.py"
  sed -i "{s/VerySecurePassphrase/$RANDOM$DATE1$RANDOM$DATE2/}" $CRYPT_LIB
  python3 -mpy_compile $CRYPT_LIB
  mkdir -p $YATCLIB
  mv ./lib/__pycache__/crypt.*.pyc $YATCLIB/crypt.pyc
  chmod 755 -R $YATCLIB

  # make user autologin
  # http://blog.shvetsov.com/2010/09/auto-login-ubuntu-user-from-cli.html
  TTY_CONF=/etc/init/tty1.conf
  sed -i '{s/exec/#exec/}' $TTY_CONF 
  echo "exec /bin/login -f $RDPUSER < /dev/tty1 > /dev/tty1 2>&1" >> $TTY_CONF

  # add some permissions to user
  echo -e "$RDPUSER\tALL=(root) NOPASSWD:/sbin/reboot,/sbin/poweroff,/bin/mount,/bin/umount\n" >> /etc/sudoers

  # setup X and application to start on boot at terminal
  BASH_PROFILE=/home/$RDPUSER/.bash_profile
  echo -e "dbus-launch --exit-with-session pulseaudio --daemon\n. startx" > $BASH_PROFILE

  # /home/user/.xinitrc
  XINITRC=/home/$RDPUSER/.xinitrc
  echo "python3 $YATCBIN $BACKEND -- -depth 32" > $XINITRC

  # fix onership
  chown $RDPUSER:$RDPUSER -R /home/$RDPUSER

  # set cupsd to listen on all interfaces
  # and allow access from subnet
  CUPSD_CONF=/etc/cups/cupsd.conf
  mv $CUPSD_CONF $CUPSD_CONF.orig
  awk '/\/Location/ {print "  Allow from 10.*"; print; next }1' $CUPSD_CONF.orig >> $CUPSD_CONF
  sed -i '{s/127.0.0.1/0.0.0.0/}' $CUPSD_CONF
  sed -i '{s/localhost/0.0.0.0/}' $CUPSD_CONF
  service cups restart

  # create directory to store mountpoints
  mkdir /media/usbdisk
  chown $RDPUSER /media/usbdisk

  # enable puppet
  puppet agent --enable

  echo -e "\tCommon configuration finished."
  return 0
}

freerdp() {
  echo -e "\tFreeRDP installation starting.\n"

  # install FreeRDP from git
  apt-get install -y build-essential git-core cmake libssl-dev libx11-dev libxext-dev libxinerama-dev \
    libxcursor-dev libxdamage-dev libxv-dev libxkbfile-dev libasound2-dev libcups2-dev libxml2 libxml2-dev \
    libxrandr-dev libgstreamer0.10-dev libgstreamer-plugins-base0.10-dev libxi-dev libavcodec-dev libpulse-dev

  cd /tmp
  git clone $FREERDP_REPO
  cd ./FreeRDP
  git checkout $FREERDP_BRANCH

  cmake -DCMAKE_BUILD_TYPE=Debug -DWITH_SSE2=ON -DWITH_CUPS=ON -DCHANNEL_PRINTER=ON -DWITH_PULSE=ON .

  make
  make install

  echo "/usr/local/lib/freerdp" > /etc/ld.so.conf.d/freerdp.conf
  ldconfig

  echo -e "\tFreeRDP installation finished."
  return 0
}


twoxclient() {
  echo -e "\t2X RDP client installation starting.\n"

  dpkg --add-architecture i386

  cd /tmp
  wget $TWOXCLIENT -O ./2xclient.deb

  dpkg -i ./2xclient.deb || apt-get -f -y install
  dpkg -i ./2xclient.deb

  echo -e "\t2X RDP client installation finished."
  return 0
}

##### BEGIN ###################################################################

if [[ $# -eq 0 ]]; then
  echo -e "\tYou must specify RDP backend [freerdp|2xclient]."
  exit 2
else
  BACKEND=$1

  echo -e "\tStarting to install YATC with $BACKEND RDP backend."
  case $BACKEND in
    freerdp)
      check_user
      common
      freerdp
    ;;
    2xclient)
      check_user
      common
      twoxclient
    ;;
    *)
      echo -e "\tUnknown backend: $BACKEND."
      exit 2
    ;;
  esac

  echo -e "\tYATC installation finished. Rebooting in 3 seconds."
  sleep 3
  reboot
fi
